Interface out AccessControlMetadataMetastore

losipiuk opened the pull request.

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftMetastoreUtil.java:
```diff
@@ -263,45 +261,29 @@ protected RoleGrant computeNext()
     }
 
-    public static boolean isRoleApplicable(SemiTransactionalHiveMetastore metastore, HivePrincipal principal, String role)
+    public static boolean isRoleApplicable(HivePrincipal principal, String role, Function<HivePrincipal, Set<RoleGrant>> listRoleGrants)
```
Maybe we pass `SqlStandardAccessControlMetadataMetastore` here, instead of a function?

although, i see the value of this too

losipiuk replied:
Yeah - it would work. Using `Function` here was kinda an intermediate step. I squash this with `Interface out SqlStandardAccessControlMetadataMetastore`

losipiuk replied:
Naa - I take that back. Let's keep the function. Going with `SqlStandardAccessControMetastore` implies that I need to also pass `ConnectoSecurityContext` in `isRoleApplicable` signature. It does not make it substantially simpler

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftMetastoreUtil.java:
```diff
@@ -263,45 +261,29 @@ protected RoleGrant computeNext()
     }
 
-    public static Stream<String> listApplicableRoles(SemiTransactionalHiveMetastore metastore, HivePrincipal principal)
+    private static Stream<String> listApplicableRoleNames(HivePrincipal principal, Function<HivePrincipal, Set<RoleGrant>> listRoleGrants)
```
here too, maybe SqlStandardAccessControlMetadataMetastore instead of a function

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftMetastoreUtil.java:
```diff
@@ -263,45 +261,29 @@ protected RoleGrant computeNext()
     }
 
-    public static Stream<HivePrincipal> listEnabledPrincipals(SemiTransactionalHiveMetastore metastore, ConnectorIdentity identity)
+    public static Stream<HivePrincipal> listEnabledPrincipals(ConnectorIdentity identity, Function<HivePrincipal, Set<RoleGrant>> listRoleGrants)
```
here too, maybe SqlStandardAccessControlMetadataMetastore instead of a function

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/security/AccessControlMetadataFactory.java:
```diff
@@ -13,9 +13,9 @@
 public interface AccessControlMetadataFactory
 {
-    AccessControlMetadata create(SemiTransactionalHiveMetastore metastore);
+    // TODO get rid of SqlStandardAccessControlMetastore. Possibly replace with something generic
+    // not tied to selected access control implementation. E.g. HiveTransactionHandle.
```
Are you sure this TODO is actionable? or it is going to stay with us for ever?

Also, since you're improving the situation here, i don't feel like you need to add a TODO at all

losipiuk replied:
I will drop it for now. Not sure if it is easily actionable. Without massive refactor.

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/security/SemiTransactionalHiveSqlStandardAccessControlMetadataMetastore.java:
```diff
@@ -0,0 +1,96 @@
+public class SemiTransactionalHiveSqlStandardAccessControlMetadataMetastore
+        implements SqlStandardAccessControlMetadataMetastore
+{
+    private final SemiTransactionalHiveMetastore hiveMetastore;
```
That might be silly, but did you consider making `SemiTransactionalHiveMetastore` implement `SqlStandardAccessControlMetadataMetastore` directly?

losipiuk replied:
Hmm - good point :)

losipiuk replied:
It may be a bit constraining if we want to change structure of it. But we can postpone adding an extra delegation layer.

findepi reviewed (commented):
First 3 commits (up to "Make ThriftMetastoreUtil not depend on SemiTransactionalHiveMetastore"):

- LGTM with commits. Can you please extract & merge (first 2 or first 3, depending on comments)

- rest - some comments, and depends what you do with the comments for the first 3, but overall LGTM too, need another closer look

losipiuk commented:
AC + rebase

losipiuk commented:
rebase

jirassimok reviewed (commented):
The code itself looks fine, but the names of the classes are far too long and non-descriptive for my tastes. There might not be good, short names for these, but I'm sure there's something better than `SemiTransactionalHiveSqlStandardAccessControlMetastore`.

If necessary, I'd move some classes into a subpackage and removing its name from theirs could help a bit (`hive` and `standardsql` are probably good candidates). Even without making a package, I think `Hive` could be removed from the names it's in (or it should be added to the others; these classes all appear to be equally Hive-related).

---

I'd also add documentation comments  the new interfaces, particularly to ensure that each one indicates what sets it apart from the others.

`SqlStandardAccessControlMetadataMetastore` in particular looks like a version of `AccessControlMetadata` that doesn't need a `ConnectorSession` to work, but I had to look at their methods one by one to figure that out, and I can't easily tell what that for their implementations. (And why is one a `Metastore` and the other not?)

losipiuk commented:
> The code itself looks fine, but the names of the classes are far too long and non-descriptive for my tastes. There might not be good, short names for these, but I'm sure there's something better than `SemiTransactionalHiveSqlStandardAccessControlMetastore`.
> 
> If necessary, I'd move some classes into a subpackage and removing its name from theirs could help a bit (`hive` and `standardsql` are probably good candidates). Even without making a package, I think `Hive` could be removed from the names it's in (or it should be added to the others; these classes all appear to be equally Hive-related).
> 
> I'd also add documentation comments the new interfaces, particularly to ensure that each one indicates what sets it apart from the others.
> 
> `SqlStandardAccessControlMetadataMetastore` in particular looks like a version of `AccessControlMetadata` that doesn't need a `ConnectorSession` to work, but I had to look at their methods one by one to figure that out, and I can't easily tell what that for their implementations. (And why is one a `Metastore` and the other not?)

I dropped `Hive` from names and added some javadoc comments.

findepi commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/security/SqlStandardAccessControlMetadata.java:
```diff
@@ -218,6 +218,14 @@ public void revokeTablePrivileges(ConnectorSession session, SchemaTableName sche
         return result.build();
     }
 
+    private Stream<HivePrincipal> getEnabledPrincipals(ConnectorIdentity identity)
```
Why `io.trino.plugin.hive.metastore.thrift.ThriftMetastoreUtil#listEnabledPrincipals` inlined here?

uninline?

losipiuk replied:
hmm - interesting. Done

jirassimok reviewed (approved):
It looks good to me.

(Though I'm still not sure we need `SemiTransactional` in the interface names.)
